### PR TITLE
LPS-143148 Send external model instead of internal model when sending webhooks notifications

### PR DESCRIPTION
**Pull request by LuismiBarcos**
Related ticket -> [LPS-143148](https://issues.liferay.com/browse/LPS-143148)

### Send external model instead of internal model

This PR substitutes the entity model that is sent to the webhook's endpoint. The ideal case is to send the model that is specified in the correspondent OpenAPI and that is sent by the Rest APIs so we are consistent with the information Liferay is sharing with external systems. Also, sharing the internal model is dangerous because it could contain sensitive information.

These are the most important changes:
* Extract logic from the `object-api` module to the `object-service` module. To avoid adding more "ugly" dependencies to the `object-api` module (like a dependency to Vulcan) a service tracker was created to register the necessary model listener. Also, the model listener was moved from the `object-api` to the `object-service` module extracting more logic from the `-api` and adding it to `-service`.
* Send the external model instead of the internal model using the DTO converter of the entity. **There are some cases where the internal model will be sent**:
  * When the DTOConverter for the entity does not exist (There isn't an API with a correspondent external model)
  * When the DTOConverter cannot convert the internal model to the external model for some reason. Normally, will be an error or unsupported case in the DTOConverter that should be fixed, so it is important to log the issue

### How to test this functionality
* Deploy the `object-api`, `object-service` and `headless-admin-user-impl` modules.
* Go to Objects and create webhook actions for an object
* Create/update/delete information of the object selected
* Check the information sent by Liferay to the webhook URL

### CODE REVIEW

**Comment by LuismiBarcos** [comment]
Remember that this PR should not be sent directly to Brian, it should be sent to the Object team so they can check the implementation. 🙂

**Comment by LuismiBarcos** [comment]
Fix SF and semantic versioning issues

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

File: modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java
Comment (javierdearcos) [inline modules/apps/object/object-api/src/main/java/com/liferay/object/system/BaseSystemObjectDefinitionMetadata.java:53]: Very nice! Much more cleaner now

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:47]: Good job generalizing 👏

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:15]: Do we want to get rid of the system part of the java package?

Other option could be to leave it as it was before with internal like `com.liferay.object.internal.system.model.listener`

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:128]: Not sure about this userId, but it is not our logic so we can leave it as it is.

What I understand is that we will use the user making the action that triggers the webhook or the userId that creates the baseModel. I think it doesn't make sense but alright.

Anyway I'll try to find a way to pass my concerns to the object team

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegister.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegister.java:1]: I think we have already a pattern for these type of classes.

Take a look at: https://github.com/liferay/liferay-portal/blob/b3173da81b62430f7150b695d5e944b224279580/modules/apps/export-import/export-import-resources-importer/src/main/java/com/liferay/exportimport/resources/importer/internal/extender/ResourceImporterExtender.java#L55

From what I see we can call ServiceTrackerFactory.create to create the tracker and declare an inner private class for the ServiceTrackerCustomizer.

I also see that there is a naming pattern to call these classes ...Extender, but I'm not sure if this applies to our use case. I bet that it could

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:115]: Although it will throw a NullPointerException I will add an explicit condition to check if the dtoConverter is null and return the internal model

Comment (javierdearcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:115]: Before this try-catch block I mean

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:130]: We can really throw a PortalException? or any kind of exception if the DTOConverter fails?

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:92]: Same doubts with userId. I will ask if this makes sense

**Review by javierdearcos — CHANGES_REQUESTED** [review]
Review my comments. The change I request mostly is to follow the pattern of creating a new ServiceTracker.

Very good job! 🎉

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (LuismiBarcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:47]: Of course is a good job. It was your idea 😁

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (LuismiBarcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:15]: Mi though was that as already exists a package for model listeners, makes sense to me to put this one there. 
Also, exists the package `internal.system` which contains the `UserSystemObjectDefinitionMetadata` and adding another package just for the model listener seems weird to me. 
But maybe is just me and makes total sense 😅

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegister.java
Comment (LuismiBarcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegister.java:1]: Perfect! I will change it to follow the pattern.

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java
Comment (LuismiBarcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:130]: I think so, because the DTOConverter's methods throw Exception. 
But this change does not come from the DTO but from the getUser method that I use in the _getExternalModel method to get the user. Because that method throws a PortalException

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java
Comment (LuismiBarcos) [reply modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/ObjectEntryModelListener.java:115]: Done!

**Comment by LuismiBarcos** [comment]
I've updated the PR following the comments:
* I've changed the implementation of the to follow the common pattern used in Liferay
* I've renamed the file `SystemObjectDefinitionMetadataRegister.java` to `SystemObjectDefinitionMetadataRegistry.java` to follow the name convention that we use in Liferay
* I've updated some tests in `object-test` that were failing because of the changes made.
* I've added a null check for the DTOConverter before trying to do the conversion

**Comment by LuismiBarcos** [comment]
ci:test:sf

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegistry.java
Comment (javierdearcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/system/SystemObjectDefinitionMetadataRegistry.java:1]: Nice! It is much cleaner like this

**Review by javierdearcos — APPROVED** [review]
👏 👏 👏

**Comment by javierdearcos** [comment]
Let's wait for the tests to finish (and if you want we can review again the description) and if everything is green we can send it to the Objects team

**Comment by LuismiBarcos** [comment]
I've just updated the description following your notes 🙂

**Comment by javierdearcos** [comment]
> I've just updated the description following your notes 🙂
Perfect now.

So let's wait for the test results and send the PR to the [objects team](https://github.com/liferay-objects) if everything is fine

**Comment by LuismiBarcos** [comment]
Ok, this is funny. I've looked at the CI errors because at first look it does not seem related, but it is related.
The class of the tests that are failing has a `@Before` method to set up some objects. At this point, everything is normal. But, one of these objects is a user. How could this affect us?

Well, by default, a User is an Object, so, when this user for test purposes is created, the `SystemObjectDefinitionMetadataModelListener` is executed. At this point, an exception is thrown because of the problems explained in this PR. To resume, when adding a user, the service adds and updates the user more than once, so, there is a point where the contact associated with the user is not set and the DTOConverter fails, and throw an exception. The thing is that this exception is handled and logged, but in the tests, this exception causes that error.

I've been testing in other modules that set a user (like `JournalArticleInfoItemPermissionProviderTest`) and it fails too.

What can we do about that? Maybe I did something wrong handling the exception? 🤔 
Did we find the first problem related to converting each entity to Object? 

Should we change the logic to get the external model and move it to the `WebhookObjectActionExecutorImpl` class? This last option came up and it was a little bit tricky because it was necessary to change some interfaces (breaking version, but not relevant) and do some changes in the classes that implement those interfaces. The problem was that some of these changes were very difficult (like the one in `ObjectActionTriggerMessageListener`) and we discard the idea.

**Comment by LuismiBarcos** [comment]
Ok, I'm a fraud. I couldn't extract the logic into a generic method. I will try again 😞 
Now I will check that at least this solution works and there is no test failing. 🤞

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by javierdearcos** [comment]
We can review this in the "Working together"

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

File: modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java
Comment (LuismiBarcos) [inline modules/apps/object/object-service/src/main/java/com/liferay/object/internal/model/listener/SystemObjectDefinitionMetadataModelListener.java:185]: I've changed the log to debug because the failing tests were related to logging the exception as an error

**Comment by LuismiBarcos** [comment]
ci:test:sf

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by LuismiBarcos** [comment]
The failing test does not seem to be related. I am not able to reproduce that error in local and there is no error in the classes of this PR. Maybe is just the CI. I will try again but from my point of view we can finally send this PR to the object team

**Comment by LuismiBarcos** [comment]
ci:test:relevant

**Comment by javierdearcos** [comment]
Green tests!!!!

And a conflict...

Rebase when you can and we can cross our fingers to pass the tests again and forward it

**Comment by LuismiBarcos** [comment]
Forward = send to object team right??
It is necessary to pass the tests again or as we have passed the tests here I can send them directly to the object team and pass the tests in that PR?

**Comment by javierdearcos** [comment]
Ohh that's true.

Sorry. Just rebase and send to Objects Team

**Comment by LuismiBarcos** [comment]
[Sent](https://github.com/liferay-objects/liferay-portal/pull/396) to the [Object team](https://github.com/liferay-objects/liferay-portal)

**Comment by LuismiBarcos** [comment]
Closing this PR as now is waiting for a review by the Object team

**Pull request closed without merging**
